packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -394,31 +394,35 @@ export class Pipeline extends PipelineBase {
 
   /** @internal */
   public _attachActionToPipeline(stage: Stage, action: IAction, actionScope: CoreConstruct): FullActionDescriptor {
-    const richAction = new RichAction(action, this);
+    const actionEnv = new ActionEnvironment(action, this);
 
     // handle cross-region actions here
-    const crossRegionInfo = this.ensureReplicationResourcesExistFor(richAction);
+    const crossRegionInfo = this.ensureReplicationResourcesExistFor(actionEnv);
 
     // get the role for the given action, handling if it's cross-account
-    const actionRole = this.getRoleForAction(stage, richAction, actionScope);
-
-    // // CodePipeline Variables
-    validateNamespaceName(richAction.actionProperties.variablesNamespace);
+    const actionRole = this.getRoleForAction(stage, actionEnv, actionScope);
 
     // bind the Action
-    const actionConfig = richAction.bind(actionScope, stage, {
-      role: actionRole ? actionRole : this.role,
+    const config = action.bind(actionScope, stage, {
+      role: actionRole ?? this.role,
       bucket: crossRegionInfo.artifactBucket,
     });
 
-    return new FullActionDescriptor({
-      // must be 'action', not 'richAction',
-      // as those are returned by the IStage.actions property,
-      // and it's important customers of Pipeline get the same instance
-      // back as they added to the pipeline
+    if (config.boundAction && config.configuration) {
+      throw new Error('The Action returned both a \'configuration\' and a \'boundAction\'; it should return one or the other');
```

Comment:
That's literally what we're doing.

Why are you opposed to validating that the implementation doesn't make a mistake? Seems like something you'd want.

Reply:
> That's literally what we're doing.

No, because you throw an exception if both are set.

> Why are you opposed to validating that the implementation doesn't make a mistake? Seems like something you'd want.

Because, as a consumer, I have no way of unblocking myself if an Action I use makes this mistake, which kind of sucks.

Reply:
> No, because you throw an exception if both are set.

Right, because the implementor does not hold to the contract. I'd want to yell about that, because there is a misapprehension going on.

> Because, as a consumer, I have no way of unblocking myself if an Action I use makes this mistake, which kind of sucks.

I'm not sure how we'd get to that situation though. No Action implemented today would trip this, and I would assume future  `Action` implementations would be unit tested (so if they implement the contract incorrectly they would not be released and you would not be using them).